js/src/admin/components/UserListPage.tsx

```diff
@@ -0,0 +1,393 @@
+import EditUserModal from '../../common/components/EditUserModal';
+import LoadingIndicator from '../../common/components/LoadingIndicator';
+import Button from '../../common/components/Button';
+
+import icon from '../../common/helpers/icon';
+import listItems from '../../common/helpers/listItems';
+
+import type User from '../../common/models/User';
+
+import ItemList from '../../common/utils/ItemList';
+import classList from '../../common/utils/classList';
+import extractText from '../../common/utils/extractText';
+
+import AdminPage from './AdminPage';
+
+type ColumnData = {
+  /**
+   * Column title
+   */
+  name: String;
+  /**
+   * Component(s) to show for this column.
+   */
+  content: (user: User) => JSX.Element;
+};
+
+type ApiPayload = {
+  data: Record<string, unknown>[];
+  included: Record<string, unknown>[];
+  links: {
+    first: string;
+    next?: string;
+  };
+};
+
+type UsersApiResponse = User[] & { payload: ApiPayload };
+
+/**
+ * Admin page which displays a paginated list of all users on the forum.
+ */
+export default class UserListPage extends AdminPage {
+  /**
+   * Number of users to load per page.
+   */
+  private numPerPage: number = 50;
+
+  /**
+   * Current page number. Zero-indexed.
+   */
+  private pageNumber: number = 0;
+
+  /**
+   * Total number of forum users.
+   *
+   * Fetched from the active `AdminApplication` (`app`), with
+   * data provided by `AdminPayload.php`, or `flarum/statistics`
+   * if installed.
+   */
+  readonly userCount: number = app.data.statistics.users.total;
+
+  /**
+   * Get total number of user pages.
+   */
+  private getTotalPageCount(): number {
+    if (this.userCount === -1) return 0;
+
+    return Math.ceil(this.userCount / this.numPerPage);
+  }
+
+  /**
+   * This page's array of users.
+   *
+   * `undefined` when page loads as no data has been fetched.
+   */
+  private pageData: User[] | undefined = undefined;
+
+  /**
+   * Are there more users available?
+   */
+  private moreData: boolean = false;
+
+  private isLoadingPage: boolean = false;
+
+  headerInfo() {
+    return {
+      className: 'UserListPage',
+      icon: 'fas fa-users',
+      title: app.translator.trans('core.admin.user_list.title'),
+      description: app.translator.trans('core.admin.user_list.description'),
+    };
+  }
+
+  /**
+   * Asynchronously fetch the next set of users to be rendered.
+   *
+   * Returns an array of Users, plus the raw API payload.
+   *
+   * Uses the `this.numPerPage` as the response limit.
+   *
+   * @param offset Offset to pass to API
+   */
+  async loadNextUserSet(offset: number = 0, pageNumber) {
```

Comment:
Couldn't we infer offset from pageNumber?

Also, `loadPage` feels like a better name